estate/models/estate_property.py

```diff
@@ -0,0 +1,82 @@
+
+
+class EstateProperty(models.Model):
+    _name = "estate.property"
+    _description = "Estate Property"
+    _order = "id desc"
+    _check_expected_price = models.Constraint("CHECK(expected_price>0)", "Le prix doit être strictement positif.")
+    _check_selling_price = models.Constraint("CHECK(selling_price>=0)", "Le prix doit être positif.")
+
+    name = fields.Char(required=True)
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date(default=fields.Date.add(fields.Date.today(), months=3), copy=False)
+    expected_price = fields.Float(required=True)
+    selling_price = fields.Float(readonly=True, copy=False)
+    bedrooms = fields.Integer(default=2)
+    living_area = fields.Integer()
+    facades = fields.Integer()
+    garage = fields.Boolean()
+    garden = fields.Boolean()
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection(
+        string='Garden orientation',
+        selection=[('North', 'N'), ('South', 'S'), ('East', 'E'), ('West', 'W')],
+        help="Specify the orientation of the garden to know when you're gonna enjoy the sun")
+    state = fields.Selection(
+        selection=[('New', 'New'), ('Offer Received', 'Offer Received'), ('Offer Accepted', 'Offer Accepted'), ('Sold', 'Sold'), ('Cancelled', 'Cancelled')],
+        default='New'
+    )
+    active = fields.Boolean(default=True)
+    salesman = fields.Many2one("res.users")
+    buyer = fields.Many2one("res.partner", copy=False)
+    tag_ids = fields.Many2many("estate.property.tag", string="Tags")
+    property_type_id = fields.Many2one("estate.property.type")
+    offer_ids = fields.One2many("estate.property.offer", "property_id")
+    total_area = fields.Float(compute="_compute_total")
+    best_price = fields.Float(compute="_compute_highest_price")
+
+    @api.depends("living_area", "garden_area")
+    def _compute_total(self):
+        for record in self:
+            record.total_area = record.living_area + record.garden_area
+
+    @api.depends("offer_ids")
+    def _compute_highest_price(self):
+        for record in self:
+            record.best_price = max(record.offer_ids.mapped("price")) if record.offer_ids else 0
+
+    @api.onchange("garden")
+    def _onchange_garden(self):
+        for record in self:
+            record.garden_area = 10 if record.garden else 0
+            record.garden_orientation = "North" if record.garden else None
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_new_or_cancelled(self):
+        for record in self:
+            if record.state not in ['New', 'Cancelled']:
+                raise UserError(_("Seules les propriétés 'New' ou 'Cancelled' peuvent être supprimées."))
+
+    def action_mark_as_sold(self):
+        for record in self:
+            if record.state == "Cancelled":
+                raise UserError(_("Cette vente a été annulée"))
+            record.state = "Sold"
+            return True
```

Comment:
Your action is used on a form, that means that you are updating only one record. So you do not need to iterate self and you can add a self.ensure_one() to make sure that your self contains only one record, if it is not the case it triggered an error, and so it protects the db. You can apply that to all method called by a button in a form.
```suggestion
    def action_mark_as_sold(self):
        self.ensure_one()
        if record.state == "Cancelled":
            raise UserError(_("Cette vente a été annulée"))
        record.state = "Sold"
        return True
```